contracts/staking/TokenStaking.sol

```diff
@@ -1315,22 +1013,87 @@ contract TokenStaking is Initializable, IStaking, Checkpoints {
+            uint96 fromAmount = authorization.authorized;
+
+            if (fromAmount <= stakingProviderStruct.tStake) {
+                continue;
+            }
+            authorization.authorized = stakingProviderStruct.tStake;
+
+            bool successful = true;
+            //slither-disable-next-line calls-loop
+            try
+                IApplication(authorizedApplication)
+                    .involuntaryAuthorizationDecrease{
+                    gas: GAS_LIMIT_AUTHORIZATION_DECREASE
+                }(stakingProvider, fromAmount, authorization.authorized)
+            {} catch {
+                successful = false;
+            }
+            if (authorization.deauthorizing > authorization.authorized) {
+                authorization.deauthorizing = authorization.authorized;
+            }
+            emit AuthorizationInvoluntaryDecreased(
+                stakingProvider,
+                authorizedApplication,
+                fromAmount,
+                authorization.authorized,
+                successful
+            );
+            if (authorization.authorized == 0) {
+                applicationsToDelete++;
+            }
+        }
+        if (applicationsToDelete > 0) {
+            cleanAuthorizedApplications(
+                stakingProviderStruct,
+                applicationsToDelete
+            );
+        }
+
+        emit Unstaked(stakingProvider, legacyStake);
+        stakingProviderStruct.keepInTStake = 0;
+        stakingProviderStruct.nuInTStake = 0;
+        decreaseStakeCheckpoint(stakingProvider, legacyStake);
+    }
+
     /// @notice Returns minimum possible stake for T, KEEP or NU in T denomination
     /// @dev For example, suppose the given staking provider has 10 T, 20 T worth
     ///      of KEEP, and 30 T worth of NU all staked, and the maximum
     ///      application authorization is 40 T, then `getMinStaked` for
     ///      that staking provider returns:
     ///          * 0 T if KEEP stake type specified i.e.
-    ///            min = 40 T max - (10 T + 30 T worth of NU) = 0 T
+    ///            min = 40 T max - (10 T) = 30 T
     ///          * 10 T if NU stake type specified i.e.
-    ///            min = 40 T max - (10 T + 20 T worth of KEEP) = 10 T
+    ///            min = 40 T max - (10 T) = 30 T
     ///          * 0 T if T stake type specified i.e.
-    ///            min = 40 T max - (20 T worth of KEEP + 30 T worth of NU) < 0 T
+    ///            min = 40 T max = 40 T < 0 T
     ///      In other words, the minimum stake amount for the specified
     ///      stake type is the minimum amount of stake of the given type
     ///      needed to satisfy the maximum application authorization given
-    ///      the staked amounts of the other stake types for that staking
-    ///      provider.
+    ///      the staked amounts of the T stake types for that staking provider.
     function getMinStaked(address stakingProvider, StakeType stakeTypes)
```

Comment:
Is there a reason to keep the `stakeType` param? Looks redundant given the changes presented in this PR.

Reply:
There are couple places to use it with Keep and Nu parameters, like `unstakeKeep`/`unstakeNu`. Mostly code to calculate authorization for all apps. I'd keep it for now just to have closer ABI to current version and I will remove `stakeType` in second PR (which will be after force unstake for all legacy stakes)